File: backend/app/schemas/cloudflare.py
"""Cloudflare domain management schemas."""

from datetime import datetime

from pydantic import BaseModel, Field


class CloudflareConfigPublic(BaseModel):
    account_id: str | None = None
    is_configured: bool
    has_api_token: bool
    has_default_dns_target: bool
    default_dns_target_type: str | None = None
    default_dns_target_value: str | None = None
    updated_at: datetime | None = None
    last_verified_at: datetime | None = None


class CloudflareConfigUpdate(BaseModel):
    account_id: str | None = Field(default=None, max_length=128)
    api_token: str | None = Field(default=None, min_length=20)
    default_dns_target_type: str | None = Field(default=None, max_length=16)
    default_dns_target_value: str | None = Field(default=None, max_length=255)


class CloudflareConnectionTestResult(BaseModel):
    success: bool
    message: str
    token_status: str | None = None


class CloudflarePageInfoPublic(BaseModel):
    page: int = 1
    per_page: int = 50
    count: int = 0
    total_count: int = 0
    total_pages: int = 1


class CloudflareZonePublic(BaseModel):
    id: str
    name: str
    status: str
    paused: bool = False
    type: str | None = None
    development_mode: int | None = None
    name_servers: list[str] = Field(default_factory=list)
    original_name_servers: list[str] = Field(default_factory=list)
    created_on: datetime | None = None
    modified_on: datetime | None = None
    activated_on: datetime | None = None


class CloudflareZonesPublic(BaseModel):
    items: list[CloudflareZonePublic]
    page_info: CloudflarePageInfoPublic


class CloudflareZoneCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    account_id: str | None = Field(default=None, max_length=128)
    jump_start: bool = False


class CloudflareDNSRecordPublic(BaseModel):
    id: str
    zone_id: str
    type: str
    name: str
    content: str
    ttl: int
    proxied: bool | None = None
    proxiable: bool | None = None
    comment: str | None = None
    priority: int | None = None
    tags: list[str] = Field(default_factory=list)
    created_on: datetime | None = None
    modified_on: datetime | None = None


class CloudflareDNSRecordsPublic(BaseModel):
    items: list[CloudflareDNSRecordPublic]
    page_info: CloudflarePageInfoPublic


class CloudflareDNSRecordCreate(BaseModel):
    type: str = Field(min_length=1, max_length=32)
    name: str = Field(min_length=1, max_length=255)
    content: str = Field(min_length=1, max_length=4096)
    ttl: int = Field(default=1, ge=1, le=86400)
    proxied: bool | None = None
    comment: str | None = Field(default=None, max_length=500)
    priority: int | None = Field(default=None, ge=0, le=65535)


class CloudflareDNSRecordUpdate(CloudflareDNSRecordCreate):
    pass


__all__ = [
    "CloudflareConfigPublic",
    "CloudflareConfigUpdate",
    "CloudflareConnectionTestResult",
    "CloudflarePageInfoPublic",
    "CloudflareZonePublic",
    "CloudflareZonesPublic",
    "CloudflareZoneCreate",
    "CloudflareDNSRecordPublic",
    "CloudflareDNSRecordsPublic",
    "CloudflareDNSRecordCreate",
    "CloudflareDNSRecordUpdate",
]

File: frontend/src/services/cloudflare.ts
import type { CancelablePromise } from "@/client"
import { OpenAPI } from "@/client"
import { request as __request } from "@/client/core/request"

export type CloudflareConfigPublic = {
  account_id: string | null
  is_configured: boolean
  has_api_token: boolean
  has_default_dns_target: boolean
  default_dns_target_type: string | null
  default_dns_target_value: string | null
  updated_at: string | null
  last_verified_at: string | null
}

export type CloudflareConfigUpdate = {
  account_id?: string | null
  api_token?: string | null
  default_dns_target_type?: string | null
  default_dns_target_value?: string | null
}

export type CloudflareConnectionTestResult = {
  success: boolean
  message: string
  token_status: string | null
}

export type CloudflarePageInfoPublic = {
  page: number
  per_page: number
  count: number
  total_count: number
  total_pages: number
}

export type CloudflareZonePublic = {
  id: string
  name: string
  status: string
  paused: boolean
  type: string | null
  development_mode: number | null
  name_servers: string[]
  original_name_servers: string[]
  created_on: string | null
  modified_on: string | null
  activated_on: string | null
}

export type CloudflareZonesPublic = {
  items: CloudflareZonePublic[]
  page_info: CloudflarePageInfoPublic
}

export type CloudflareZoneCreate = {
  name: string
  account_id?: string | null
  jump_start: boolean
}

export type CloudflareDNSRecordPublic = {
  id: string
  zone_id: string
  type: string
  name: string
  content: string
  ttl: number
  proxied: boolean | null
  proxiable: boolean | null
  comment: string | null
  priority: number | null
  tags: string[]
  created_on: string | null
  modified_on: string | null
}

export type CloudflareDNSRecordsPublic = {
  items: CloudflareDNSRecordPublic[]
  page_info: CloudflarePageInfoPublic
}

export type CloudflareDNSRecordMutation = {
  type: string
  name: string
  content: string
  ttl: number
  proxied?: boolean
  comment?: string
  priority?: number
}

export type ZoneListParams = {
  page?: number
  per_page?: number
  search?: string
  status?: string
}

export type DNSRecordListParams = {
  page?: number
  per_page?: number
  search?: string
  type?: string
  proxied?: boolean
}

export const CloudflareApiService = {
  getConfig(): CancelablePromise<CloudflareConfigPublic> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/cloudflare/config",
    })
  },

  updateConfig(
    data: CloudflareConfigUpdate,
  ): CancelablePromise<CloudflareConfigPublic> {
    return __request(OpenAPI, {
      method: "PUT",
      url: "/api/v1/cloudflare/config",
      body: data,
      mediaType: "application/json",
    })
  },

  testConfig(): CancelablePromise<CloudflareConnectionTestResult> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/cloudflare/config/test",
    })
  },

  listZones(
    params: ZoneListParams = {},
  ): CancelablePromise<CloudflareZonesPublic> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/cloudflare/zones",
      query: params,
    })
  },

  getZone(zoneId: string): CancelablePromise<CloudflareZonePublic> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/cloudflare/zones/{zone_id}",
      path: { zone_id: zoneId },
    })
  },

  createZone(
    data: CloudflareZoneCreate,
  ): CancelablePromise<CloudflareZonePublic> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/cloudflare/zones",
      body: data,
      mediaType: "application/json",
    })
  },

  listDnsRecords(
    zoneId: string,
    params: DNSRecordListParams = {},
  ): CancelablePromise<CloudflareDNSRecordsPublic> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/cloudflare/zones/{zone_id}/dns-records",
      path: { zone_id: zoneId },
      query: params,
    })
  },

  createDnsRecord(
    zoneId: string,
    data: CloudflareDNSRecordMutation,
  ): CancelablePromise<CloudflareDNSRecordPublic> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/cloudflare/zones/{zone_id}/dns-records",
      path: { zone_id: zoneId },
      body: data,
      mediaType: "application/json",
    })
  },

  updateDnsRecord(
    zoneId: string,
    recordId: string,
    data: CloudflareDNSRecordMutation,
  ): CancelablePromise<CloudflareDNSRecordPublic> {
    return __request(OpenAPI, {
      method: "PATCH",
      url: "/api/v1/cloudflare/zones/{zone_id}/dns-records/{record_id}",
      path: { zone_id: zoneId, record_id: recordId },
      body: data,
      mediaType: "application/json",
    })
  },

  deleteDnsRecord(
    zoneId: string,
    recordId: string,
  ): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "DELETE",
      url: "/api/v1/cloudflare/zones/{zone_id}/dns-records/{record_id}",
      path: { zone_id: zoneId, record_id: recordId },
    })
  },
}
